src/Vim.re

```diff
@@ -260,100 +298,110 @@ let _getDefaultCursors = (cursors: list(Cursor.t)) =>
-          Native.vimInput(pair.closing);
-          Native.vimInput("<LEFT>");
+          let canCloseBefore = () =>
+            AutoClosingPairs.canCloseBefore(
+              line,
+              location.column,
+              autoClosingPairs,
+            );
+
+          if (v == "<BS>"
+              && AutoClosingPairs.isBetweenDeletionPairs(
+                   line,
+                   location.column,
+                   autoClosingPairs,
+                 )) {
+            Native.vimInput("<DEL>");
+            Native.vimInput("<BS>");
+          } else if (v == "<CR>" && isBetweenClosingPairs()) {
+            Native.vimInput("<CR>");
+            Native.vimInput("<CR>");
+            Native.vimInput("<UP>");
+            Native.vimInput("<TAB>");
+          } else if (AutoClosingPairs.isPassThrough(
+                       v,
+                       line,
+                       location.column,
+                       autoClosingPairs,
+                     )) {
+            Native.vimInput("<RIGHT>");
+          } else if (AutoClosingPairs.isOpeningPair(v, autoClosingPairs)
+                     && canCloseBefore()) {
+            let pair = AutoClosingPairs.getByOpeningPair(v, autoClosingPairs);
+            Native.vimInput(v);
+            Native.vimInput(pair.closing);
+            Native.vimInput("<LEFT>");
+          } else {
+            Native.vimInput(v);
+          };
         } else {
           Native.vimInput(v);
         };
-      } else {
-        Native.vimInput(v);
+        Cursor.get();
       };
-      Cursor.get();
-    };
-
-    let mode = Mode.getCurrent();
-    let cursors = _getDefaultCursors(cursors);
-    if (mode == Types.Insert) {
-      // Run first command, verify we don't go back to normal mode
-      switch (cursors) {
-      | [hd, ...tail] =>
-        let newHead = runCursor(hd);
-
-        let newMode = Mode.getCurrent();
-        // If we're still in insert mode, run the command for all the rest of the characters too
-        let remainingCursors =
-          if (newMode == Types.Insert) {
-            List.map(runCursor, tail);
-          } else {
-            tail;
-          };
 
-        [newHead, ...remainingCursors];
-      // This should never happen...
-      | [] => cursors
-      };
-    } else {
-      switch (cursors) {
-      | [hd, ..._] => Cursor.set(hd)
-      | _ => ()
+      let mode = Mode.getCurrent();
+      let cursors = _getDefaultCursors(cursors);
+      if (mode == Types.Insert) {
+        // Run first command, verify we don't go back to normal mode
+        switch (cursors) {
+        | [hd, ...tail] =>
+          let newHead = runCursor(hd);
+
+          let newMode = Mode.getCurrent();
+          // If we're still in insert mode, run the command for all the rest of the characters too
+          let remainingCursors =
+            if (newMode == Types.Insert) {
+              List.map(runCursor, tail);
+            } else {
+              tail;
+            };
```

Comment:
A switch tends to be better for enums, as you get the benefit of exhaustiveness checking if you enumerate all the constructors. It's also more convenient if you need to change the behaviour for certain modes.